### PR TITLE
Tb ibmdbpy changes

### PR DESCRIPTION
**Pull request by ToniBollinger**
The changes fix issues https://github.com/ibmdbanalytics/ibmdbpy/issues/53 and https://github.com/ibmdbanalytics/ibmdbpy/issues/55. 
In addition it fixes SQL code generation problems related to lower case column names.

New testcases have been added to cover these changes. All are executed successfully:

```
============================= test session starts ==============================
platform linux -- Python 3.7.2, pytest-5.3.5, py-1.8.1, pluggy-0.13.1 -- /usr/local/bin/python3.7
cachedir: .pytest_cache
rootdir: /mnt/bludata0/scratch/ibmdbpy
plugins: flaky-3.6.1
collecting ... collected 492 items
test_aggregation.py::Test_aggregation::test_idadf_aggregation_1_column_add_int PASSED [  0%]
test_aggregation.py::Test_aggregation::test_idadf_aggregation_1_column_mul_int PASSED [  0%]
test_aggregation.py::Test_aggregation::test_idadf_aggregation_1_column_div_int PASSED [  0%]
test_aggregation.py::Test_aggregation::test_idadf_aggregation_1_column_sub_int PASSED [  0%]
test_aggregation.py::Test_aggregation::test_idadf_aggregation_1_column_floordiv_int PASSED [  1%]

....

test_utils.py::Test_utils::test_check_viewname PASSED                    [ 99%]
test_utils.py::Test_utils::test_check_case PASSED                        [ 99%]
test_utils.py::Test_utils::test_reset_attributes PASSED                  [100%]

=============================== warnings summary ===============================
ibmdbpy/tests/test_aggregation.py::Test_aggregation::test_idadf_aggregation_1_column_add_int
ibmdbpy/tests/test_base_private.py::Test_IdaDataBase_PrivateMethods::test_idadb_get_valid_tablename_error
ibmdbpy/tests/test_base_private.py::Test_IdaDataBase_PrivateMethods::test_idadb_get_valid_viewname_error
ibmdbpy/tests/test_base_private.py::Test_IdaDataBase_PrivateMethods::test_idadb_get_valid_modelname_error
ibmdbpy/tests/test_base_table_manipulation.py::Test_DeleteDataBaseObjects::test_idadb_drop_view
ibmdbpy/tests/test_base_table_manipulation.py::Test_TableManipulation::test_idadb_rename_value_error
ibmdbpy/tests/test_utils.py::Test_utils::test_reset_attributes
  /usr/local/lib/python3.7/site-packages/ibmdbpy/utils.py:238: UserWarning: Mixed case names are not supported in database object names.
    warnings.warn("Mixed case names are not supported in database object names.", UserWarning)

ibmdbpy/tests/test_frame.py::Test_IdaDataFrameBehavior::test_idadf_getitem_slice
ibmdbpy/tests/test_frame.py::Test_IdaDataFrameBehavior::test_idaseries_getitem_slice
  /usr/local/lib/python3.7/site-packages/ibmdbpy/indexing.py:117: UserWarning: Row order is not guaranteed if no indexer was given and the dataset was not sorted
    warnings.warn("Row order is not guaranteed if no indexer" +

-- Docs: https://docs.pytest.org/en/latest/warnings.html
===Flaky Test Report===

test_idadb_exists_model_negative passed 1 out of the required 1 times. Success!

===End Flaky Test Report===
====== 487 passed, 1 skipped, 4 xpassed, 9 warnings in 201.52s (0:03:21) =======
```

### CODE REVIEW

File: ibmdbpy/tests/test_base_connexion.py [hunk 1]
```diff
@@ -34,10 +34,7 @@ def test_idadb_attr(self, idadb):
         assert isinstance(idadb.data_source_name, six.string_types)
         assert isinstance(idadb._con_type, six.string_types)
         assert isinstance(idadb._idadfs, list)
-        if idadb._con_type == 'odbc':
-            assert isinstance(idadb._connection_string, six.string_types)
-        elif idadb._con_type == 'jdbc':
-            assert isinstance(idadb._connection_string, list)
+        assert isinstance(idadb._connection_string, six.string_types)
```
Comment (mineo) [inline ibmdbpy/tests/test_base_connexion.py]: Please revert this change, it doesn't work for JDBC connections:

```
self = <test_base_connexion.Test_ConnectToDB object at 0x7f75f81609e8>
idadb = <ibmdbpy.base.IdaDataBase object at 0x7f76af645470>

    def test_idadb_attr(self, idadb):
        assert isinstance(idadb.data_source_name, six.string_types)
        assert isinstance(idadb._con_type, six.string_types)
        assert isinstance(idadb._idadfs, list)
>       assert isinstance(idadb._connection_string, six.string_types)
E       AssertionError: assert False
E        +  where False = isinstance(['jdbc:db2://<hostname>:50000/BLUDB', 'bluadmin', '<password>'], (<class 'str'>,))
E        +    where ['jdbc:db2://<hostname>:50000/BLUDB', 'bluadmin', '<password>'] = <ibmdbpy.base.IdaDataBase object at 0x7f76af645470>._connection_string
E        +    and   (<class 'str'>,) = six.string_types

test_base_connexion.py:37: AssertionError
```

File: ibmdbpy/geoFrame.py [hunk 2]
```diff
@@ -1028,22 +1028,9 @@ def _binary_operation_handler(self, ida2, db2gse_function,
...
-        arguments_for_db2gse_function = []
-        if 'IDA1' or 'DB2GSE' in column1_for_db2gse:
-            arguments_for_db2gse_function.append(column1_for_db2gse) 
-        else:
-            arguments_for_db2gse_function.append('IDA1.'+column1_for_db2gse)
-        if 'IDA2' or 'DB2GSE' in column2_for_db2gse:
-            arguments_for_db2gse_function.append(column2_for_db2gse)
-        else:
-            arguments_for_db2gse_function.append('IDA2.'+column2_for_db2gse)
+        column1_for_db2gse = 'IDA1.\"%s\"' %(ida1.geometry.column)
```
Comment (mineo) [inline ibmdbpy/geoFrame.py]: Please clarify this change. This is now basically the code from before eva's changes in https://github.com/ibmdbanalytics/ibmdbpy/pull/50/files#diff-5624c304e87b1a80d0a03491cd424a55, which, as I understand, did not work correctly in all cases. I admit that that's far better than not working in any case, which we have now, but she was trying to solve a problem with that change.

File: ibmdbpy/base.py [hunk 3]
```diff
@@ -565,22 +565,29 @@ def exists_model(self, modelname):
...
+
+        data = self.ida_query("call idax.list_models('schema=%s, where=MODELNAME=''%s''')" % (modelschema, modelname))
 
         if not data.empty:
-            if modelname in data['MODELNAME'].values:
-                return True
+            return True
 
         tablelist = self.show_tables(show_all=True)
+        tablelist= tablelist[(tablelist['TABSCHEMA']==modelschema) & (tablelist['TABNAME']==modelname)]
```
Comment (mineo) [inline ibmdbpy/base.py]:
```suggestion
        tablelist = tablelist[(tablelist['TABSCHEMA']==modelschema) & (tablelist['TABNAME']==modelname)]
```

File: ibmdbpy/base.py [hunk 4]
```diff
@@ -565,22 +565,29 @@ def exists_model(self, modelname):
...
         if len(tablelist):
-            if modelname in tablelist['TABNAME'].values:
-                tabletype = tablelist[tablelist['TABNAME'] == modelname]['TYPE'].values[0]
-                raise TypeError("%s exists but is not a model (of type '%s')"
-                                %(modelname, tabletype))
-            else:
-                return False
+             tabletype = tablelist['TYPE'].values[0]
+             raise TypeError("%s.%s exists, but is not a model (of type '%s')"
+                  %(modelschema, modelname, tabletype))
```
Comment (mineo) [inline ibmdbpy/base.py]: Please keep the alignment for readability:

```suggestion
             raise TypeError("%s.%s exists, but is not a model (of type '%s')"
                             % (modelschema, modelname, tabletype))
```

File: ibmdbpy/statistics.py [hunk 5]
```diff
@@ -637,8 +637,8 @@ def cov(idadf, other = None):
         raise TypeError("cov() missing 1 required positional argument: 'other'")
 
     columns = idadf._get_numerical_columns()
-    if not columns:
-        print(idadf.name + " has no numeric columns")
+    if not columns or len(columns) < 2:
```
Comment (mineo) [inline ibmdbpy/statistics.py]: Please remove the 3 tests marked `xfail` in https://github.com/ibmdbanalytics/ibmdbpy/blob/master/ibmdbpy/tests/test_statistics.py#L190-L193. They are now not failing by accident, but by design.

File: ibmdbpy/geoFrame.py (hunk 2 above)
Comment (ToniBollinger) [reply ibmdbpy/geoFrame.py]: The main difference is that before the change the value of column1_for_db2gse (and column2_for_db2gse)  can be an expression like 'DB2GSE.ST_MBR("SHAPE")'. After the change it is the name of a column, the geometry column, like 'MBR'. This suffices because the association between the defining expression and the column name is included in the SQL statement of the name attribute, for instance:
``
'(SELECT * FROM (SELECT "OBJECTID","SHAPE","STATEFP","COUNTYFP","COUNTYNS","NAME","GEOID","NAMELSAD","LSAD","CLASSFP","MTFCC","CSAFP","CBSAFP","METDIVFP","FUNCSTAT","ALAND","AWATER","INTPTLAT","INTPTLON",DB2GSE.ST_MBR("SHAPE") AS "MBR" FROM SAMPLES.GEO_COUNTY) WHERE ("NAME" = \'Austin\'))'
``

File: ibmdbpy/statistics.py (hunk 5 above)
Comment (ToniBollinger) [reply ibmdbpy/statistics.py]: They succeeded:
``
test_statistics.py::Test_DescriptiveStatistics::test_idadf_statistics_one_column[cov] XPASS [ 96%]
test_statistics.py::Test_DescriptiveStatistics::test_idadf_statistics_one_column[corr] XPASS [ 96%]
test_statistics.py::Test_DescriptiveStatistics::test_idadf_statistics_one_column[quantile] PASSED [ 96%]
test_statistics.py::Test_DescriptiveStatistics::test_idadf_statistics_one_column[mad] XPASS [ 96%]
``
So why should we remove them?

Comment (mineo) [reply ibmdbpy/statistics.py]: Oh, right, the only indication of an error in cov/corr/mad is the message that's being printed. That makes the tests pretty much useless, but means you can at least remove the `xfail` marker from them. That will change their result from `XPASS` (expected to fail, but passed) to just `PASS`

File: ibmdbpy/tests/test_base_connexion.py (hunk 1 above)
Comment (ToniBollinger) [reply ibmdbpy/tests/test_base_connexion.py]: The type of the connection string can be both - a list or a string - depending on the jaydebeapi version. I've modified the testcase to accept both types for jdbc connection strings with commit https://github.com/ibmdbanalytics/ibmdbpy/pull/56/commits/5be5189df0f278f1aff012f45b3ca7bfbaa69554

File: ibmdbpy/learn/association_rules.py [hunk 6]
```diff
@@ -541,23 +541,19 @@ def _retrieve_AssociationRules_Model(self, modelname, verbose=False):
...
 
-        assocpatterns = self._idadb.ida_query('SELECT * FROM "' +
-        self._idadb.current_schema + '"."' + modelname + '_ASSOCPATTERNS"')
+        assocpatterns = self._idadb.ida_query('SELECT * FROM ' + modelname + '_ASSOCPATTERNS')
         assocpatterns.columns = ["ITEMSETID","ITEMID"]
         assocpatterns.columns = [x.upper() for x in assocpatterns.columns]
 
-        assocpatterns_stats = self._idadb.ida_query('SELECT * FROM "' +
-        self._idadb.current_schema + '"."' + modelname + '_ASSOCPATTERNS_STATISTICS"')
+        assocpatterns_stats = self._idadb.ida_query('SELECT * FROM ' + modelname + '_ASSOCPATTERNS_STATISTICS')
```
[the remark this answers is not in the document]
Comment (ToniBollinger) [reply ibmdbpy/learn/association_rules.py]: Model names can include only alphanumeric characters and underscores because they have to pass the checks of  ibmddbpy.utils.check_tablename which turns lowercase to uppercase characters, 
for instance see
https://github.com/ibmdbanalytics/ibmdbpy/blob/7019ae3829874e800782ea398ec400f4dce3664e/ibmdbpy/learn/association_rules.py#L229-L235
Therefore it is not necessary to include the model schema and model name inside double quotes.

File: ibmdbpy/base.py [hunk 7]
```diff
@@ -565,22 +565,29 @@ def exists_model(self, modelname):
         """
         modelname = ibmdbpy.utils.check_tablename(modelname)
         if '.' in modelname:
+            modelschema = modelname.split('.')[0]
```
[the remark this answers is not in the document]
Comment (ToniBollinger) [reply ibmdbpy/base.py]: No, this is done by ibmdbpy.utils.check_tablename. 
But I have added a testcase for mixed case model names (see  commit https://github.com/ibmdbanalytics/ibmdbpy/pull/56/commits/cb66560e79fc6ddc9d604d4822fec19f37b26fd8)

File: ibmdbpy/tests/test_base.py [hunk 8]
```diff
@@ -90,12 +91,27 @@ def test_idadb_exists_model_positive(self, idadb, idadf_tmp):
             idadb.drop_model(kmeans.modelname)
         except : pass
 
+    def test_idadb_exists_model_with_schema_positive(self, idadb, idadf_tmp):
+        idadb.add_column_id(idadf_tmp, destructive=True)
+        # Create a simple KMEANS model
+        kmeans = KMeans(n_clusters=3, modelname="MYSCHEMA.MODEL_58979457385")
+        kmeans.fit(idadf_tmp)
+        assert(idadb.exists_model("MYSCHEMA.MODEL_58979457385") == 1)
```
[the remark this answers is not in the document]
Comment (ToniBollinger) [reply ibmdbpy/tests/test_base.py]: Changed with commit https://github.com/ibmdbanalytics/ibmdbpy/pull/56/commits/cb66560e79fc6ddc9d604d4822fec19f37b26fd8

File: ibmdbpy/statistics.py (hunk 5 above)
Comment (ToniBollinger) [reply ibmdbpy/statistics.py]: Changed with commit https://github.com/ibmdbanalytics/ibmdbpy/pull/56/commits/6d8043c019bcf089e107474b3f698a0c3ba976f3

File: ibmdbpy/learn/association_rules.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mineo) [reply ibmdbpy/learn/association_rules.py]: Let's not focus on upper and lower case for now, what about model names containing spaces? I suppose that's supported, at least I couldn't find any documentation claiming otherwise (for example https://www.ibm.com/support/knowledgecenter/SS6NHC/com.ibm.swg.im.dashdb.analytics.doc/doc/r_model_names_examples.html doesn't add any restrictions, https://www.ibm.com/support/knowledgecenter/SS6NHC/com.ibm.swg.im.dashdb.analytics.doc/doc/r_model_component_naming_conventions.html even talks about a model called `Sample Model`).

File: ibmdbpy/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (mineo) [reply ibmdbpy/base.py]: Thank you!

File: ibmdbpy/learn/association_rules.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (ToniBollinger) [reply ibmdbpy/learn/association_rules.py]: idmdbpy is more restrictive concerning model names than the analytic stored procedures. The same is true, by the way, for table names.
It supports only modelnames with alphanumeric characters and underscores. 
This checked is performed by` ibmdbpy.utils.check_tablename(modelname)`, for instance for exists_model in:
https://github.com/ibmdbanalytics/ibmdbpy/blob/7019ae3829874e800782ea398ec400f4dce3664e/ibmdbpy/base.py#L539-L541
...
https://github.com/ibmdbanalytics/ibmdbpy/blob/7019ae3829874e800782ea398ec400f4dce3664e/ibmdbpy/base.py#L564-L566

So to make that restriction clear I would describe it in the ibmdbpy documentation, like it is done in the docstring for IdaDataBase.rename:
https://github.com/ibmdbanalytics/ibmdbpy/blob/7019ae3829874e800782ea398ec400f4dce3664e/ibmdbpy/base.py#L1026-L1038

[the remark this answers is not in the document]
Comment (mineo) [reply ibmdbpy/learn/association_rules.py]: Thanks for digging up that information, looks like the point about whitespace characters is moot.

File: ibmdbpy/geoFrame.py (hunk 2 above)
Comment (mineo) [reply ibmdbpy/geoFrame.py]: It's still not entirely clear to me how this works, but it's better than the current situation :-D
